### PR TITLE
[cache] Use provided nix bin cache from api

### PR DESCRIPTION
**Pull request by mikeland73**
## Summary

* If user is logged in, check jetpack API to see if they have nix cache available. If available use it.
* Adds new provider framework for logged-in/logged-out functionality.

Note: There's a minor issue we need to fix with the access token. It's missing the audience which is needed by the API. Will fix in https://github.com/jetpack-io/opensource

## How was it tested?

* Copied over access token with correct audience
* Added myself as a trusted user
* `devbox add hello` on org that has nix cache. Observed it attempted to use the cache.

### CODE REVIEW

File: internal/devbox/providers/nixcache/nixcache.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	"go.jetpack.io/devbox/internal/devbox/providers/identity"
+	"go.jetpack.io/devbox/internal/nix"
+	"go.jetpack.io/devbox/internal/ux"
+	"go.jetpack.io/pkg/api"
+	"go.jetpack.io/pkg/auth"
+)
+
+type Provider struct{}
+
+var singleton *Provider = &Provider{}
```
Comment (mohsenari) [inline internal/devbox/providers/nixcache/nixcache.go:20]: why do providers need to be singletons?

File: internal/devbox/providers/nixcache/nixcache.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	// * Automate making user trusted if needed
+	if !nix.IsUserTrusted(ctx) {
+		ux.Fwarning(
+			os.Stderr,
+			"In order to use a custom nix cache you must be a trusted user. Please "+
+				"add your username to nix.conf (usually located at /etc/nix/nix.conf)"+
+				" and restart the nix daemon.",
+		)
+	}
+}
```
Comment (mohsenari) [inline internal/devbox/providers/nixcache/nixcache.go:99]: if you're just showing a warning when a user is not trusted then `ensureTrustedUser` is not the best name for it. 
Maybe `checkTrustedUser` or alternatively we can return `NixCacheConfig{}` if the user is not trusted.

File: internal/devbox/secrets.go [hunk 3]
```diff
@@ -18,7 +17,6 @@ func (d *Devbox) UninitializedSecrets(ctx context.Context) *envsec.Envsec {
 		},
 		IsDev:      build.IsDev,
 		Stderr:     d.stderr,
-		Store:      &jetstore.JetpackAPIStore{},
```
Comment (mohsenari) [inline internal/devbox/secrets.go:21]: is this change relevant to the PR?

File: internal/nix/build.go [hunk 4]
```diff
@@ -29,6 +30,7 @@ func Build(ctx context.Context, args *BuildArgs, installables ...string) error {
 		cmd.Args = append(cmd.Args, "--extra-substituters", args.ExtraSubstituter)
 	}
 	cmd.Env = allowUnfreeEnv(os.Environ())
+	cmd.Env = append(cmd.Env, args.Env...)
```
Comment (mohsenari) [inline internal/nix/build.go:33]: this can be `cmd.Env = append(allowUnfreeEnv(os.Environ()), args.Env...)`

**Review by mohsenari — APPROVED** [review]
see a couple of comments

File: internal/devbox/providers/nixcache/nixcache.go (hunk 1 above)
Comment (mikeland73) [reply internal/devbox/providers/nixcache/nixcache.go:20]: They don't right now, but I'm hoping to add caching to avoid making multiple API calls. Will do this after testing end to end and understanding where the perf bottlenecks are.

File: internal/devbox/providers/nixcache/nixcache.go (hunk 2 above)
Comment (mikeland73) [reply internal/devbox/providers/nixcache/nixcache.go:99]: I can change the function name. The issue is that I'm not 100% confident we can catch all cases where you are actually allowed to use the substitutor, so returning an empty cache config would be wrong.

Especially since `nix build` will still work in the bad case (it just shows a warning)

File: internal/devbox/secrets.go (hunk 3 above)
Comment (mikeland73) [reply internal/devbox/secrets.go:21]: This is no longer used when we updated envsec dependency.
